snow/engine/common/engine.go

```diff
@@ -501,3 +514,24 @@ type InternalHandler interface {
 	// Notify this engine of a message from the virtual machine.
 	Notify(context.Context, Message) error
 }
+
+// AppError is an application-defined error
+type AppError struct {
+	// Code is application-defined and should be used for error matching
+	Code uint32
```

Comment:
`Code == 0` implies no error, right?

Reply:
`Code == 0` is an undefined error. This exists because i wanted to pass the `error` interface into the `VM#AppRequestFailed` function instead of using `AppError` concrete type. In rpcchainvm we have to parse the error to propagate it to the underlying VM, and if that fails (which it never should because rpcchainvm is an internal vm and we only pass it `AppError`) we fall back to the undefined error type which is also the zero-value of `AppError{}`. 

We could avoid this by passing the concrete `AppError` type instead of the `error` interface

```
func (vm *VMClient) CrossChainAppRequestFailed(ctx context.Context, chainID ids.ID, requestID uint32, err error) error {
	msg := &vmpb.CrossChainAppRequestFailedMsg{
		ChainId:      chainID[:],
		RequestId:    requestID,
		ErrorCode:    common.ErrUnexpected.Code,
		ErrorMessage: common.ErrUnexpected.Message,
	}

	if appErr, ok := err.(*common.AppError); ok {
		msg.ErrorCode = appErr.Code
		msg.ErrorMessage = appErr.Message
	}

	_, err = vm.client.CrossChainAppRequestFailed(ctx, msg)
	return err
}
```

Reply:
I see, might be worth documenting since usually `Code == 0` --> there was an error.

Reply:
I feel like passing in the struct directly would be pretty reasonable

Reply:
I went back and forth on this and I feel like it might be better to just use the `error` interface mostly because so much error handling stuff in go like the errors package relies on the invariant that an nil error has a nil type and a nil value which isn't true if we return a concrete type and that's passed into a function that takes in a `error` (say `require.Error` for example, or `errors.Is`).

Reply:
update: We ended up going back to the concrete type, but using the `error` interface in the SDK.